hbase-server/src/main/java/org/apache/hadoop/hbase/master/procedure/ReopenTableRegionsProcedure.java

```diff
@@ -139,33 +170,57 @@ protected Flow executeFromState(MasterProcedureEnv env, ReopenTableRegionsState
       case REOPEN_TABLE_REGIONS_CONFIRM_REOPENED:
         regions = regions.stream().map(env.getAssignmentManager().getRegionStates()::checkReopened)
           .filter(l -> l != null).collect(Collectors.toList());
-        if (regions.isEmpty()) {
-          return Flow.NO_MORE_STATE;
+        // we need to create a set of region names because the HRegionLocation hashcode is only
+        // based
+        // on the server name
+        Set<byte[]> currentRegionBatchNames = currentRegionBatch.stream()
+          .map(r -> r.getRegion().getRegionName()).collect(Collectors.toSet());
+        currentRegionBatch = regions.stream()
+          .filter(r -> currentRegionBatchNames.contains(r.getRegion().getRegionName()))
+          .collect(Collectors.toList());
+        if (currentRegionBatch.isEmpty()) {
+          if (regions.isEmpty()) {
+            return Flow.NO_MORE_STATE;
+          } else {
+            setNextState(ReopenTableRegionsState.REOPEN_TABLE_REGIONS_REOPEN_REGIONS);
+            if (reopenBatchBackoffMillis > 0) {
+              backoff(reopenBatchBackoffMillis);
+            }
+            return Flow.HAS_MORE_STATE;
+          }
         }
-        if (regions.stream().anyMatch(loc -> canSchedule(env, loc))) {
+        if (currentRegionBatch.stream().anyMatch(loc -> canSchedule(env, loc))) {
           retryCounter = null;
           setNextState(ReopenTableRegionsState.REOPEN_TABLE_REGIONS_REOPEN_REGIONS);
+          if (reopenBatchBackoffMillis > 0) {
+            backoff(reopenBatchBackoffMillis);
+          }
           return Flow.HAS_MORE_STATE;
         }
         // We can not schedule TRSP for all the regions need to reopen, wait for a while and retry
         // again.
         if (retryCounter == null) {
           retryCounter = ProcedureUtil.createRetryCounter(env.getMasterConfiguration());
         }
-        long backoff = retryCounter.getBackoffTimeAndIncrementAttempts();
+        long backoffMillis = retryCounter.getBackoffTimeAndIncrementAttempts();
         LOG.info(
-          "There are still {} region(s) which need to be reopened for table {} are in "
+          "There are still {} region(s) which need to be reopened for table {}. {} are in "
             + "OPENING state, suspend {}secs and try again later",
-          regions.size(), tableName, backoff / 1000);
-        setTimeout(Math.toIntExact(backoff));
-        setState(ProcedureProtos.ProcedureState.WAITING_TIMEOUT);
-        skipPersistence();
+          regions.size(), tableName, currentRegionBatch.size(), backoffMillis / 1000);
+        backoff(backoffMillis);
         throw new ProcedureSuspendedException();
       default:
         throw new UnsupportedOperationException("unhandled state=" + state);
     }
   }
 
+  private void backoff(long millis) throws ProcedureSuspendedException {
+    setTimeout(Math.toIntExact(millis));
+    setState(ProcedureProtos.ProcedureState.WAITING_TIMEOUT);
+    skipPersistence();
```

Comment:
This might need more research. I know other examples skip persistence, besides they make no change. But in this case maybe we need persistence so that we save an updated list of regions still needing to reopen

Reply:
I think it is OK to skip persistence here as we do not persist the reopen of the region in procedure state, IIRC we will use the openSeqNum to determine whether the region has already been reopened.
But the problem here is that, the old logic is for error retrying, where we can not schedule TRSP for some regions, but here, we are doing throttling, so I do not think we should keep increase the retry number and increase the retry interval while scheduling TRSPs...

Reply:
In what event do we rely on the persisted state? Do we rely on it when moving between states, or is it only relevant when recovering from an hmaster roll or something? I agree with @bbeaudreault that we don't reevaluate the `regions` in `REOPEN_TABLE_REGIONS_REOPEN_REGIONS`, so if we rely on the persisted state to decide which regions to transition after a batch backoff then I think this would stick us in an infinite loop. e.g.,:
1. `this.regions = [a, b, c, d, e]`, `batchSize=1`, `batchBackoffMs=1`
2. enter `..._REOPEN_REGIONS` with `this.regions = [a, b, c, d, e]`, derive currentBatch of `a` and kick off TRSP
3. enter `..._CONFIRM_REOPENED`, evaluate `this.regions = [b, c, d, e]` because a has reopened, acknowledge the success by setting current state to `WAITING_TIMEOUT`, next state back to `..._REOPEN_REGIONS` & throw `ProcedureSuspendedException` without persistence of the new this.regions value
4. `..._REOPEN_REGIONS` would be loaded from the persisted state with `this.regions = [a, b, c, d, e]`, so we'd be back at step 2 and destined to repeat this cycle

To avoid this we could just persist our state if the current batch has no regions still in transition. All this said, I feel like the above example misunderstands how proc state is used because otherwise I don't see how [this test](https://github.com/apache/hbase/blob/a3a880439fa553f0b96973bb051884c4a4998ea7/hbase-server/src/test/java/org/apache/hadoop/hbase/master/procedure/TestReopenTableRegionsProcedureBatchBackoff.java#L74-L88) would pass. It succeeds which pretty clearly indicates the lack of an infinite loop

re: 
> I do not think we should keep increase the retry number and increase the retry interval while scheduling TRSPs

We will only increase the retry interval and backoff if our current batch has yet to successfully reopen — the interval will remain fixed and will not increment retry counts if we're successfully moving onto the next batch of regions

Reply:
It is always OK to persist the state, no problem. Skipping persistent is for optimizing the performance, as there is no actual state change when retrying. That's why I say it should be OK to skip persistent here as the reopened flag is not stored in the procedure state.
On the retrying part, checked the code again, you just use a fixed backoff timeout to call the backoff method so it should be OK.

Reply:
I think we need some way to track which regions still have to be restarted after backoff. I don't think we're modifying the proto here yet so agree persistence wouldn't help much yet. But maybe we need to update the proto?

Reply:
Ya so the current region list is serialized into the proto. I think we need to not skip persistence and update the serialization to either remove the current batch or add some sort of offset to the proto. Otherwise let's say this crashed right before finishing -- restarting up, it's have to reopen all of the regions again (in our case often many thousands). If we persist an offset we can pick up right where we left off.

Reply:
Ok thanks @Apache9 for the helpful context. I just re-read the code, and now I see the checkReopened. I wasn't aware of that, good to know!

In that case, as you say, we don't need persistence. So we can move on to your other feedback on final fields, etc. Thanks!